File: 901.cpp
#include <vector>

using namespace std;

class StockSpanner {
private:
  vector<int> vals;
  vector<int> spans;

public:
  StockSpanner() {}

  int next(int price) {
    if (vals.empty() || vals.back() > price) {
      vals.push_back(price);
      spans.push_back(1);
      return 1;
    }
    int span = 1;
    int i = vals.size() - 1;
    while (i > -1 && vals[i] <= price) {
      span += spans[i];
      i -= spans[i];
    }
    vals.push_back(price);
    spans.push_back(span);
    return span;
  }
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */

File: 901.py
class StockSpanner:
    def __init__(self):
        self.vals = []
        self.spans = []

    def next(self, price: int) -> int:
        if not self.vals or self.vals[-1] > price:
            self.vals.append(price)
            self.spans.append(1)
            return 1
        i = len(self.vals) - 1
        span = 1
        while i > -1 and self.vals[i] <= price:
            span += self.spans[i]
            i -= self.spans[i]
        self.spans.append(span)
        self.vals.append(price)
        return span


# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)

obj = StockSpanner()
obj.next(100)
obj.next(80)
obj.next(60)
obj.next(70)
obj.next(60)
obj.next(75)
obj.next(85)
print(obj.spans)
print("expected: ")
print([1, 1, 1, 2, 1, 4, 6])
